optuna/pruners/threshold.py

```diff
@@ -0,0 +1,113 @@
+import math
+from typing import Optional
+
+import optuna
+from optuna.pruners import BasePruner
+from optuna.pruners.percentile import _is_first_in_interval_step
+
+
+class ThresholdPruner(BasePruner):
+    """Pruner to detect outlying metrics of the trials.
+
+    Prune if a metric exceeds upper bound threshold or
+    falls behind lower bound threshold.
+
+    Example:
+        .. testcode::
+
+            from optuna import create_study
+            from optuna.exceptions import TrialPruned
+            from optuna.pruners import ThresholdPruner
+
+
+            def objective_1(trial):
+                for step in range(n_trial_step):
+                    trial.report(ys_for_upper[step], step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+
+
+            def objective_2(trial):
+                for step in range(n_trial_step):
+                    trial.report(ys_for_lower[step], step)
+
+                    if trial.should_prune():
+                        raise TrialPruned()
+
+
+            ys_for_upper = [0.0, 0.1, 0.2, 0.5, 1.2]
+            ys_for_lower = [100.0, 90.0, 0.1, 0.0, -1]
+            n_trial_step = 5
+
+            study = create_study(pruner=ThresholdPruner(upper=1.0))
+            study.optimize(objective_1, n_trials=10)
+
+            study = create_study(pruner=ThresholdPruner(lower=0.0))
+            study.optimize(objective_2, n_trials=10)
+
+
+    Args
+        lower:
+            minimum value which determines whether pruner prunes or not
+            (If value is smaller than lower, it prunes)
+        upper:
+            maximum value which determines whether pruner prunes or not
+            (If value is larger than upper, it prunes)
+        n_warmup_steps:
+            Pruning is disabled until the trial exceeds the given number of step.
+        interval_steps:
+            Interval in number of steps between the pruning checks, offset by the warmup steps.
+            If no value has been reported at the time of a pruning check, that particular check
+            will be postponed until a value is reported. Value must be at least 1.
+
+    """
+
+    def __init__(
+            self,
+            lower: Optional[float] = None,
+            upper: Optional[float] = None,
+            n_warmup_steps: int = 0,
+            interval_steps: int = 1
+    ) -> None:
```

Comment:
In the case `lower == upper == None`, `ThresholdPruner` should raise an Exception since

1. we can create an instance of `ThresholdPruner` doesn't prune anything by setting `lower = -inf, upper = inf`
2. we can create a pruner that never prune a training using `NopPruner`.